language-support/hs/bindings/src/DA/Ledger.hs

```diff
@@ -108,102 +112,84 @@ mdm :: MetadataMap
 mdm = MetadataMap Map.empty
 
 
-
 ----------------------------------------------------------------------
--- Services with streaming responses
-
-data Elem a = Elem a | Eend | Eerr String
-
-deElem :: Elem a -> IO a
-deElem = \case
-    Elem a -> return a
-    Eend -> fail "readStream, end"
-    Eerr s -> fail $ "readStream, err: " <> s
-
-newtype Stream a = Stream { mv :: MVar (Elem a) }
-
-newStream :: IO (Stream a)
-newStream = do
-    mv <- newEmptyMVar
-    return Stream{mv}
-
-writeStream :: Stream a -> Elem a -> IO () -- internal use only
-writeStream Stream{mv} elem = putMVar mv elem
-
-takeStream :: Stream a -> IO a
-takeStream Stream{mv} = takeMVar mv >>= deElem
-
-data StreamState = SS -- TODO
-getStreamContents :: Stream a -> IO ([a],StreamState)
-getStreamContents Stream{mv} = do xs <- loop ; return (xs,SS)
-    where
-        loop = do
-            tryTakeMVar mv >>= \case
-                Nothing -> return []
-                Just e -> do
-                    x <- deElem e
-                    xs <- loop
-                    return (x:xs)
-
--- wrap LL.Transaction to show summary
-newtype LL_Transaction = LL_Transaction { low :: LL.Transaction } --TODO: remove
-    deriving Eq
-
-instance Show LL_Transaction where
-    show LL_Transaction{low} = _summary
-        where
-            _summary = "Trans:id=" <> Text.unpack transactionTransactionId
-            _full = show low
-            LL.Transaction{transactionTransactionId} = low
-
-
--- TODO: return (HL) [Transaction]
-transactions :: LedgerHandle -> Party -> IO (Stream LL_Transaction)
-transactions LedgerHandle{port,lid} party = wrapE "transactions" $ do
+-- transaction_service
+
+getLedgerEnd :: LedgerHandle -> IO LedgerOffset
+getLedgerEnd LedgerHandle{port,lid} = wrapE "getLedgerEnd" $ do
+    LL.withGRPCClient (config port) $ \client -> do
+        service <- LL.transactionServiceClient client
+        let TransactionService _ _ _ _ _ _ rpc = service
```

Comment:
Using the record selector for the last field would be more appropriate here.

Reply:
so: `TransactionService{transactionServiceGetLedgerEnd=rpc}`
Although this is differently fragile to changes in the proto generated code.
Not fragile on additional or reorded rpcs; but now fragile on the rpc name.
But overall this seems better.

Reply:
I would write
```haskell
let rpc = transactionServiceGetLedgerEnd service
```